R/manage.R

```diff
@@ -0,0 +1,804 @@
+  max_version <- numeric_version(paste(max_version, collapse = "."))
+
+  min_version <- rep(0, 4)
+  min_version[seq_along(version_parts)] <- version_parts
+  min_version <- numeric_version(paste(min_version, collapse = "."))
+
+  available_versions <- available_versions[
+    available_versions <= max_version &
+      available_versions >= min_version
+  ]
+
+  if (length(available_versions) == 0) {
+    return(NULL)
+  }
+
+  max(available_versions)
+}
+
+curl_fetch_headers <- function(url) {
+  h <- curl::new_handle()
+  curl::handle_setopt(h, nobody = TRUE)
+  req <- curl::curl_fetch_memory(url, handle = h)
+  req_parse_headers(req)
+}
+
+req_parse_headers <- function(req) {
+  headers <- rawToChar(req$headers)
+  parsed_headers <- strsplit(headers, "\r\n")[[1]]
+  parsed_headers <- parsed_headers[parsed_headers != ""]
+  parsed_headers <- strsplit(parsed_headers, ": ")
+  parsed_headers <- rlang::set_names(
+    lapply(parsed_headers, `[`, 2),
+    sapply(parsed_headers, `[`, 1)
+  )
+
+  parsed_headers
+}
+
+req_headers_last_modified <- function(headers) {
+  names(headers) <- tolower(names(headers))
+  withr::with_locale(new = c("LC_TIME" = "en_US"), {
+    last_modified <- as.POSIXct(
+      headers[["last-modified"]],
+      format = "%a, %d %b %Y %H:%M:%S GMT",
+      tz = "GMT"
+    )
+    last_modified
+  })
+}
+
+download_json_cached <- function(url, update_cached = TRUE) {
+  path_cache <- chromote_cache_path()
+  dir.create(path_cache, showWarnings = FALSE, recursive = TRUE)
+
+  path_local <- file.path(path_cache, basename(url))
+
+  # Check if local file exists and get its modified time
+  if (file.exists(path_local)) {
+    if (!update_cached) {
+      return(path_local)
+    }
+
+    local_mtime <- file.info(path_local)$mtime
+
+    is_local_stale <- tryCatch(
+      {
+        # Fetch headers from the server
+        headers <- curl_fetch_headers(url)
+        server_last_modified <- req_headers_last_modified(headers)
+
+        length(server_last_modified) == 1 && local_mtime < server_last_modified
+      },
+      error = function(err) {
+        rlang::inform(
+          "Could not reach Chrome for Testing to update available versions.",
+          parent = err
+        )
+        FALSE
+      }
+    )
+
+    # Compare local file time with server's last-modified
+    if (is_local_stale) {
+      # message("Source URL not modified, using cached version")
+      return(path_local)
+    }
+  }
+
+  req <- curl::curl_fetch_memory(url)
+
+  if (req$status_code == 200) {
+    # message("Source URL was updated, downloading new content")
+    json_content <- rawToChar(req$content)
+    writeLines(json_content, path_local)
+
+    # Set the local file's modified time to the last-modified
+    last_modified <- req_headers_last_modified(req_parse_headers(req))
+    Sys.setFileTime(path_local, last_modified)
+  } else {
+    stop("Could not download ", url, ": Status code ", req$status_code)
+  }
```

Comment:
Cosmetic.

Also changed `stop()` to `cli::cli_abort()`

```suggestion
  if (req$status_code != 200) {
    cli::cli_abort("Could not download {.val {url}}: Status code {.val {req$status_code}}")
  }

  # message("Source URL was updated, downloading new content")
  json_content <- rawToChar(req$content)
  writeLines(json_content, path_local)

  # Set the local file's modified time to the last-modified
  last_modified <- req_headers_last_modified(req_parse_headers(req))
  Sys.setFileTime(path_local, last_modified)
```

Reply:
I'll never turn down an early return :)

Reply:
I'll have to make this change locally. If we're using `cli::cli_abort()` we should include the response in the error condition.

Reply:
Done in c1081491154442cbdf9aace362d5358d468a63f9